payjoin/src/send/error.rs

```diff
@@ -235,3 +236,179 @@ impl std::error::Error for CreateRequestError {
+                },
+            Err(e) => return e.into(),
+        }
+    }
+    /// Parse a response from the receiver.
+    ///
+    /// response must be valid JSON string.
+    pub fn from_str(response: &str) -> Self {
+        if let Some(parsed) = serde_json::from_str(response).ok() {
+            Self::from_json(parsed)
+        } else {
+            InternalValidationError::Parse.into()
+        }
+    }
+}
+
+impl std::error::Error for ResponseError {}
+
+impl From<WellKnownError> for ResponseError {
+    fn from(value: WellKnownError) -> Self { Self::WellKnown(value) }
+}
+
+impl From<InternalValidationError> for ResponseError {
+    fn from(value: InternalValidationError) -> Self {
+        Self::Validation(ValidationError { internal: value })
+    }
+}
+
+// It is imperative to carefully display pre-defined messages to end users and the details in debug.
+impl Display for ResponseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::WellKnown(e) => e.fmt(f),
+            // Don't display unknowns to end users, only debug logs
+            Self::Unrecognized(_, _) => write!(f, "The receiver sent an unrecognized error."),
+            Self::Validation(e) => write!(f, "The receiver sent an invalid response: {}", e),
+        }
+    }
+}
+
+impl fmt::Debug for ResponseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::WellKnown(e) => write!(
+                f,
+                r#"Well known error: {{ "errorCode": "{}",
+                "message": "{}" }}"#,
+                e.error_code(),
+                e.to_string(),
+            ),
+            Self::Unrecognized(code, msg) => write!(
+                f,
+                r#"Unrecognized error: {{ "errorCode": "{}", "message": "{}" }}"#,
+                code, msg
+            ),
+            Self::Validation(e) => write!(f, "Validation({:?})", e),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum WellKnownError {
+    Unavailable,
+    NotEnoughMoney,
+    VersionUnsupported,
+    OriginalPsbtRejected,
+}
+
+impl WellKnownError {
+    pub fn error_code(self) -> &'static str {
+        match self {
+            WellKnownError::Unavailable => "unavailable",
+            WellKnownError::NotEnoughMoney => "not-enough-money",
+            WellKnownError::VersionUnsupported => "version-unsupported",
+            WellKnownError::OriginalPsbtRejected => "original-psbt-rejected",
+        }
+    }
+}
+
+impl FromStr for WellKnownError {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, ()> {
+        match s {
+            "unavailable" => Ok(WellKnownError::Unavailable),
+            "not-enough-money" => Ok(WellKnownError::NotEnoughMoney),
+            "version-unsupported" => Ok(WellKnownError::VersionUnsupported),
+            "original-psbt-rejected" => Ok(WellKnownError::OriginalPsbtRejected),
+            _ => Err(()),
+        }
```

Comment:
Yes, we want wellknown error to only match on `errorCode`, but it should still contain a `message` which may be different than the template. However, only template message will be displayed to the user and the actual message printed to debug log